Assignment2/Assignment 2.py

```diff
@@ -0,0 +1,366 @@
+# 
+# Note that this is still considered to be linear model as the coefficients/weights associated with the features are still linear. x<sup>2</sup> is only a feature. However the curve that we would be fitting in this case is quadratic in nature.
+# 
+# Below we show an example of a quadratic curve being fit to the data
+
+# In[6]:
+
+
+# Create polynomial features with degree 2
+polynomial_features = PolynomialFeatures(degree=2)
+x_poly = polynomial_features.fit_transform(train_x)
+
+# Apply linear regression
+model = LinearRegression()
+model.fit(x_poly, train_y)
+y_poly_pred = model.predict(x_poly)
+
+# Plot the results
+plt.scatter(train_x, train_y, s=10)
+plt.plot(train_x, y_poly_pred, color='r')
+plt.show()
+
+
+# As you can see, we get a slightly better fit with a quadratic curve. Let's use the model to make predictions on our validation set and compute the mean squared error, which is the error which we wish to minimize.
+
+# In[7]:
+
+
+# Make predictions using pretrained model
+valid_y_poly_pred = model.predict(polynomial_features.fit_transform(valid_x))
+
+# Calculate root mean squared error
+mse = mean_squared_error(valid_y, valid_y_poly_pred)
+print("Mean Squared Error: {}".format(mse))
+
+# Plot the prediction results
+plt.scatter(valid_x, valid_y, s=10)
+plt.plot(valid_x, valid_y_poly_pred, color='r')
+plt.show()
+
+
+# ## Question 1: Polynomial Regression Using Sci-Kit Learn
+# 
+# Now it is your turn! Following the same format as above, implement a 10-degree polynomial regression model on the training data and plot your results. Use your model to predict the output of the validation set and calculate the mean square error. Report and plot the results. 
+
+# In[15]:
+
+
+### YOUR CODE HERE - Fit a 10-degree polynomial using Sci-Kit Learn
+polynomial_features = PolynomialFeatures(degree = 10)
+x_poly_10 = polynomial_features.fit_transform(train_x)
+
+### YOUR CODE HERE - Plot your the curve on the training data set
+plt.scatter(train_x, train_y, s = 10)
+plt.plot(train_x, y_poly_pred_10, color = 'r')
+plt.show()
+
+### YOUR CODE HERE - Use model to predict output of validation set
+model = LinearRegression()
+model.fit(x_poly_10, train_y)
+y_poly_pred_10 = model.predict(x_poly_10)
+output = model.predict(polynomial_features.fit_transform(valid_x))
+
+### YOUR CODE HERE - Calculate the RMSE. Report and plot the curve on the validation set.
+RMSE = mean_squared_error(valid_y, output)
+print("RMSE: {}".format(RMSE))
+
+plt.scatter(valid_x, valid_y, s = 10)
+plt.plot(valid_x, output, color = 'r')
+plt.show()
+
+
+# #### Did the root mean squared error go up or down as compared to the 2-degree polynomial curve? Why do you think this is the case?
+# 
+# ------- ANSWER HERE -----------
+# 
+# The RMSE goes down compared to the 2-degree polynomial curve. This can be the case that data is better represented by a 10-degree polynomial than a quadratic polynomial.
```

Comment:
YES! you can also say that the 2-degree polynomial curve would underfit the data :)